src/controllers/hid/hidiothread.cpp

```diff
@@ -0,0 +1,330 @@
+    emit receive(QByteArray(reinterpret_cast<const char*>(pCurrentBuffer),
+                         bytesRead),
+            mixxx::Time::elapsed());
+}
+
+QByteArray HidIoThread::getInputReport(unsigned int reportID) {
+    auto startOfHidGetInputReport = mixxx::Time::elapsed();
+    auto lock = lockMutex(&m_hidDeviceMutex);
+
+    m_pPollData[m_pollingBufferIndex][0] = reportID;
+    // FIXME: implement upstream for hidraw backend on Linux
+    // https://github.com/libusb/hidapi/issues/259
+    int bytesRead = hid_get_input_report(
+            m_pHidDevice, m_pPollData[m_pollingBufferIndex], kBufferSize);
+    if (bytesRead <= kReportIdSize) {
+        // -1 is the only error value according to hidapi documentation.
+        // Otherwise minimum possible value is 1, because 1 byte is for the reportID,
+        // the smallest report with data is therefore 2 bytes
+        qCWarning(m_logInput)
+                << "getInputReport is unable to get data from"
+                << m_deviceInfo.formatName() << "serial #"
+                << m_deviceInfo.serialNumber() << ":"
+                << mixxx::convertWCStringToQString(
+                           hid_error(m_pHidDevice), kMaxHidErrorMessageSize);
+        // Note, that the GetInputReport request is optional, according to the HID specification,
+        // not all devices implement it.
+        return QByteArray();
+    }
+
+    // Convert array of bytes read in a JavaScript compatible return type, this is returned as deep-copy, for thread safety.
+    // For compatibility with HidController::processInputReport, the reportID prefix is included added here
+    QByteArray returnArray = QByteArray(
+            reinterpret_cast<char*>(m_pPollData[m_pollingBufferIndex]),
+            bytesRead);
+
+    lock.unlock();
+
+    qCDebug(m_logInput) << bytesRead << "bytes received by hid_get_input_report"
+                        << m_deviceInfo.formatName() << "serial #"
+                        << m_deviceInfo.serialNumber()
+                        << "(including one byte for the report ID:"
+                        << QString::number(static_cast<quint8>(reportID), 16)
+                                   .toUpper()
+                                   .rightJustified(2, QChar('0'))
+                        << ") - Needed: "
+                        << (mixxx::Time::elapsed() - startOfHidGetInputReport)
+                                   .formatMicrosWithUnit();
+
+    return returnArray;
+}
+
+void HidIoThread::cacheOutputReport(const QByteArray& data, unsigned int reportID) {
+    {
+        auto lock = lockMutex(&m_outputReportMapMutex);
+        if (m_outputReports.find(reportID) == m_outputReports.end()) {
+            std::unique_ptr<HidIoReport> pNewOutputReport;
+            m_outputReports[reportID] = std::make_unique<HidIoReport>(
+                    reportID, data.size());
+        }
+    }
+    m_outputReports[reportID]->cacheOutputReport(data, m_deviceInfo, m_logOutput);
```

Comment:
We must use an iterator here instead of invoking operator[] twice, which is a (potentially) mutating operation. Otherwise the thread-safety assumptions are violated! The iterator must be obtained while the map is still locked!

You also don't need to introduce a separate scope, just use lock.unlock() before submitting the report.